File: apps/api/src/application/ports/import-job-repository.ts
import type { ImportJob, ImportJobSummary, ImportRowReport } from '@/domain/import/import-job'

export interface ProductUpsert {
  readonly line: number
  readonly sku: string
  readonly name: string
  readonly description?: string | null
  readonly price: number
  readonly stock: number
  readonly weightKg?: number | null
  readonly category?: string | null
}

export interface ImportPlan {
  readonly fileName: string
  readonly rejected: ImportRowReport[]
  readonly writes: ProductUpsert[]
}

export interface ImportJobRepository {
  commit(plan: ImportPlan): Promise<ImportJob>
  findById(id: string): Promise<ImportJob | null>
  findAll(): Promise<ImportJobSummary[]>
}

export const IMPORT_JOB_REPOSITORY = Symbol('ImportJobRepository')

File: apps/api/src/ecommerce_api/application/ports/import_job_repository.py
from dataclasses import dataclass
from decimal import Decimal
from typing import Protocol

from ecommerce_api.domain.import_job import ImportJob, ImportJobSummary, ImportRowReport


class Absent:
    def __repr__(self) -> str:
        return 'ABSENT'


ABSENT = Absent()


@dataclass(frozen=True, slots=True)
class ProductUpsert:
    line: int
    sku: str
    name: str
    price: Decimal
    stock: int
    description: str | Absent | None = ABSENT
    weight_kg: Decimal | Absent | None = ABSENT
    category: str | Absent | None = ABSENT


@dataclass(frozen=True, slots=True)
class ImportPlan:
    file_name: str
    rejected: list[ImportRowReport]
    writes: list[ProductUpsert]


class ImportJobRepository(Protocol):
    async def commit(self, plan: ImportPlan) -> ImportJob: ...

    async def find_by_id(self, id: str) -> ImportJob | None: ...

    async def find_all(self) -> list[ImportJobSummary]: ...
